[3.x] Editor Android Port

m4gr3d opened the pull request:
`3.x` version of the Godot Editor Android port based on the work done by @thebestnom in https://github.com/godotengine/godot/pull/36776!

This PR cleans a lot of the logic in https://github.com/godotengine/godot/pull/36776 and adapt the code for the `3.x` branch. Given the stability of the `3.x` branch, this should help accelerate testing and feedback on this implementation.

There are still a lot of tweaks, UI cleanup and touch-ups needed to be done before the Editor Android port can be considered suitable for public release and usage. Nonetheless, the current logic is stable enough and partitioned away that we can start reviewing and landing this PR and then iterate on top of it.

Addresses https://github.com/godotengine/godot-proposals/issues/3931


https://user-images.githubusercontent.com/914968/152773076-9565fc2e-7079-4b39-8fa7-526de6e3eb16.mp4



### Apk Download (only meant for testing and feedback!)
[Latest 3.x Godot Editor Android build (all architectures)](https://drive.google.com/drive/folders/1Tzhu3yZFnOKwmLUCRBdbMMMHn79e_MmL?usp=sharing)

### Generate apk from sources
- Using the regular Godot build process, generate Godot tools android shared libraries for each platform you want the apk to support:
  - e.g: `scons tools=true platform=android target=<build_target> android_arch=<android_arch>`
  - `release_debug` is recommended for the `target` option
  - `android_arch` should be one of `arm64v8`, `armv7`, `x86` or `x86_64`
- Navigate to the `platform/android/java` directory
- Run `./gradlew generateGodotEditor`
  - The command will generate one apk for each available target
  - The generated apk(s) can be found in the `bin` directory.

<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

Commit:
Android port of the Godot Editor

These set of changes focus primarily on getting the core logic and overall Godot Editor UI and functionality up and running natively on Android devices.
UI tweaks / cleanup / polish, as well configuration for Android specific functionality / restrictions will be addressed in follow-up PRs iteratively based on feedback.

Co-authored-by: thebestnom <shoval.arad@gmail.com>

ElhamAryanpur commented:
This is AWESOME!

Calinou commented:
I gave the APK a try on a Samsung Galaxy ZFold2, some comments:

- Buttons are quite small, but it's usable given the large screen size (on the main display).
- The splash screen doesn't appear when I open a project in the editor – I just get a black screen for a few seconds.
- I think the default orientation should allow both landscape modes (standard and reverse).
- There could be an editor setting to allow using any orientation, including portrait mode. On tablets, portrait mode should still be wide enough to allow using the editor.
- There could be an editor setting to make the editor use fullscreen mode to maximize the available screen real estate. In fact, we may want to make this the default given most mobile displays are quite small.

Touch improvements that could be implemented in the future:

- Double-tapping in the project list and FileDialogs should work as it does on PC.
- Long-holding a node in the Scene tree dock should pop up the context menu.
- The back button currently doesn't do anything when you are in the editor or project manager. We should probably make it do something, such as exiting to the project list (with the usual confirmation).
  - To improve productivity on the go, there could be an editor setting that allows customizing the back button action while in the editor:
    - Simulate an Undo press (<kbd>Ctrl + Z</kbd>), or Redo (<kbd>Ctrl + Shift + Z</kbd>) if held for a long time.
    - Switch between the 2D/3D and Script tabs (depending on the last tab), to avoid having to tap a small icon at the top every so often.

LightningAA commented:
Agree with everything Calinou said. First thing I did was change the editor scale to 3 and soft lock myself out of the editor (can't open any projects):
![Screenshot_20220208-132200](https://user-images.githubusercontent.com/50304111/153069344-504836f6-79c4-4571-bab3-04a4f46f51ee.png)

m4gr3d commented:
@Calinou @LightningAA Thanks for the feedback!!

I'll create a set of issues to track the feedbacks so we can start addressing it.

@akien-mga Given the updates to this PR are partitioned away and should not affect the stable code, we can start reviewing and merging it. 
Then we'll address the reported feedback in separate PRs. These would prevent this one from growing too large.

thebestnom commented:
> @Calinou @LightningAA Thanks for the feedback!!
> 
> I'll create a set of issues to track the feedbacks so we can start addressing it.
> 
> @akien-mga Given the changes in this PR are partitioned away and should not affect the stable code/logic, we can start reviewing and merging it. 
> Then we'll address the reported feedback in separate PRs. These would prevent this one from growing too large.

Also any changes you did here I would glad you add back (directly without PR) to 4.0 and merge after

m4gr3d commented:
> Also any changes you did here I would glad you add back (directly without PR) to 4.0 and merge after

@thebestnom Sounds good. I'll push the changes for the `master` branch as well once this PR is reviewed and merged.

filipworksdev commented:
Some of the issues Calinou raised could be fixed by 
- Disabling features from Editor->Manage Editor Features... so you have less components at the same time.  
- Using an editor theme specifically designed for Android that have larger and more square controls that are easier to click on
- Also this PR https://github.com/godotengine/godot/pull/40287 could prove useful for an Android port 
- (Offtopic for this PR but deserves mention) Editor should remove all hardcoded minimum panel sizes to prevent overflows on smaller screens.
- Also worth mentioning there are addons that allow touch gesture which could be ported to C++.

fire-forge commented:
This is great! To add on to what everyone else has said, I have a few suggestions for this:
* Use different theme settings by default, with the editor scale increased for easier use on small screens and Additional Spacing increased to make buttons easier to touch.
* Make the area behind the navigation bar use the editor theme color instead of solid black.
* Add new editor settings for showing/hiding navigation bar and status bar (those with phones would want them disabled for more screen space, those with larger tablets might want them enabled).

fire-forge commented:
I tried it out on my phone, and I was able to run the project manager, but after making a project and pressing edit I got this screen for a few seconds and then it crashed. Not sure if that's the fault of this app or if it's because my phone is a potato. 

![Screenshot_20220209-090712](https://user-images.githubusercontent.com/67974470/153231739-652f2620-bafa-4186-866c-e2c192bd5944.png)

My phone is a Moto g7 with Android 10.

Calinou commented:
> Use different theme settings by default, with the editor scale increased for easier use on small screens and Additional Spacing increased to make buttons easier to touch.

The best approach would be to [rely on the OS-provided scaling factor](https://github.com/godotengine/godot-proposals/issues/2661). However, this scaling would often be too large to be usable, so it would have to be clamped depending on the screen's lowest dimension (width or height).

filipworksdev commented:
An interesting approach would be for Godot to provide it's own basic built in keyboard that goes in the bottom panel since most of those panels are read only with exception for the Shader Script. Obviously it won't be as effective as the system touch keyboard but it would be an integrated alternative that embeds with the interface.  There are some mobile addons that provide custom keyboard

FancyTextLabel commented:
Obviously it won't be as effective as the system touch keyboard but it would be an integrated alternative that embeds with the interface.

A more compact, simple and independent keyboard, with a switch 🔄 to change to normal keyboard!

Half screen.
![Screenshot_20220209-162753](https://user-images.githubusercontent.com/99336670/153275806-cebf8bd1-3443-4269-b997-aa225ec286ef.png)

Calinou commented:
> Obviously it won't be as effective as the system touch keyboard but it would be an integrated alternative that embeds with the interface.

We can't resize the editor window to let the virtual keyboard display, as the screen is too small to do that.

Deozaan commented:
I just tried this on a 10 inch tablet. It was pretty amazing! I also got stuck indefinitely on a black screen when trying to open a GLES3 project, but GLES2 worked after a few seconds of black screen, which I now realize was probably the duration of the splash screen.

+1 to all of Calinou's feedback, except:

> Dragging a finger on the 2D/3D viewport should pan it. Pinching with two fingers should zoom. (What would be used to orbit in the 3D viewport?)

My instinct was to drag a finger to "look around" (orbit?) in the 3D viewport. I suggest two-finger dragging for panning and pinching for zooming, as those two motions feel more closely related. It seems more intuitive to me. That said, I don't feel strongly about this and would be willing to do it Calinou's suggested way.

m4gr3d commented:
I've added a `master` (vulkan) version of the editor Android port in https://github.com/godotengine/godot/pull/58160 for those brave enough to test!

filipworksdev commented:
I have just tested it on a 10.5 inch Galaxy Tab S6 Lite. Works pretty well! 

Some issues that were not raised: 
- Playing the project (F5) will replace the editor with the running project and there is no way to stop and return to the editor. You need to restart the editor.  Android back should return from playing a project back to the editor.  
- The keyboard takes 50% of the entire editor and randomly shows up at inconvenient times. There needs to be a button  or a setting to control when or how the Android keyboard pops up. Even better, Godot should have it's own built in keyboard.
- No gestures for 2D or 3D viewports. My suggestions:
  * two finger pinch to zoom
  * single finger tap to select one
  * single finger drag to select multiple
  * two fingers drag to orbit

akien-mga commented on editor/editor_node.cpp:
```diff
@@ -472,6 +472,9 @@ void EditorNode::_notification(int p_what) {
 			get_tree()->get_root()->set_as_audio_listener(false);
 			get_tree()->get_root()->set_as_audio_listener_2d(false);
 			get_tree()->set_auto_accept_quit(false);
+#if defined(ANDROID_ENABLED) && defined(TOOLS_ENABLED)
```
All of `editor/` is only built with `TOOLS_ENABLED`, so it's redundant to specify it here too.

akien-mga commented on platform/android/java/build.gradle:
```diff
@@ -138,6 +143,10 @@ def templateBuildTasks() {
 
     // Only build the apks and aar files for which we have native shared libraries.
     for (String target : supportedTargets) {
+        if (target == "release_debug") {
+            // 'release_debug' is not supported for generating templates.
+            continue
+        }
```
There seems to be a inconsistency between the Gradle and SCons targets.

"Debug" template *are* built with `scons tools=no target=release_debug`. The `target=debug` platform is only used for actual engine debugging (i.e. getting dev asserts, more information on crashes, etc.), and we don't ship any binaries built with `target=debug` (this is only for custom builds for people who know they need it).

I see that `platform/android/detect.py` doesn't seem to do anything special for `target=release_debug` compared to `target=release`, which is a discrepancy with all other platforms (note that it's `startswith`, applies to both `release` and `release_debug`): https://github.com/godotengine/godot/blob/f5b9cbaff6f0a058d08187b3124948ae68848cd0/platform/android/detect.py#L189

akien-mga replied:
> I see that `platform/android/detect.py` doesn't seem to do anything special for `target=release_debug` compared to `target=release`, which is a discrepancy with all other platforms (note that it's `startswith`, applies to both `release` and `release_debug`):

Actually that part seems to be intentional, the difference is made in `SConstruct` directly for all platforms, seems like we factored this out at some point. Other platforms like Linux still make a difference between the two in their `detect.py` but it's just to switch from `-O3` to `-O2` between `release` and `release_debug` (and on Android we seem to only use `-O2` so it's not needed for now).

m4gr3d replied:
@akien-mga I have pushed a second (separate) commit that attempts to resolve the inconsistency by having the gradle's `buildtypes` match exactly the scons' targets.

This has the side effect of affecting the naming of the generated template apks (`android_debug.apk` is now `android_release_debug.apk`). I think that's a safe compromise to make since the template apks are not directly exposed to the end user.

akien-mga commented on platform/android/java/build.gradle:
```diff
@@ -163,6 +172,56 @@ def isAndroidStudio() {
+    include('android_editor_debug.apk')
+}
+
+task copyEditorRelease_debugBinaryToBin(type: Copy) {
+    dependsOn ':editor:assembleRelease_debug'
+    from('editor/build/outputs/apk/release_debug')
+    into(binDir)
+    include('android_editor_release_debug.apk')
+}
+
+task copyEditorReleaseBinaryToBin(type: Copy) {
+    dependsOn ':editor:assembleRelease'
+    from('editor/build/outputs/apk/release')
+    into(binDir)
+    include('android_editor_release.apk')
+}
```
That doesn't match the build matrix for other platforms.

The official released (production) editor binaries use `target=release_debug`. The editor can't be used with `target=release` only, as debugging tools are then not included, and it would crash on errors instead of reporting them.

So the only two editor targets you should support are:
`tools=yes target=release_debug` -> production build, official releases
`tools=yes target=debug` -> development build, used by engine developers (slow and not meant for use in production)

m4gr3d replied:
I have updated the logic to exclude the `release` target for editor builds and added a comment to explain the exclusion.

madmiraal replied:
We need to update the [documentation](https://docs.godotengine.org/en/stable/development/compiling/compiling_for_android.html#building-the-export-templates) for building export templates, because the default is now to build the editor version not the templates.

akien-mga replied:
Ah that default should be fixed IMO, I would expect it to still default to builtin templates.

I think it would still be the default for `scons` since there is:
```
def get_flags():
    return [
        ("tools", False),
    ]
```
which changes the default value for `tools` (but it can still be overridden with `tools=yes`).

So I guess the problem happens at the gradle step?

madmiraal replied:
Okay. So what's changed then is that the default target "debug" now builds a `dev.aar` instead of a `debug.aar`.

FYI: Previously I was using the following to build a quick test debug export template for android x86 emulators:
```
scons -j8 platform=android android_arch=x86
```
And then I run `./gradlew generateDevTemplate`
I now need to use:
```
scons -j8 target=release_debug platform=android android_arch=x86
```

Which matches the documentation, but seems less intuitive. I would expect `release_debug` to be for the editor builds  i.e. with `tools=yes`. If we're changing behaviour, it may be a good opportunity to change the names to match outputs and expectations.

akien-mga replied:
For `generateDevTemplate` is should be using `target=debug`, this might be a bug. `target=debug` is only used for development purposes and doesn't match the "Debug Template" that we build with `target=release_debug` (and which is the one `generateGodotTemplates` should use).

So it might have been an oversight among the other buildsystem changes.

> If we're changing behaviour, it may be a good opportunity to change the names to match outputs and expectations.

For `3.x` I think we should aim not to change behavior. For `master` I plan to implement https://github.com/godotengine/godot-proposals/issues/3371 which should help clean this up for Android too.

akien-mga commented on platform/android/java/build.gradle:
```diff
@@ -163,6 +172,56 @@ def isAndroidStudio() {
+    def tasks = []
+
+    for (String target : supportedTargets) {
+        if (target == "release") {
+            // Skip release for now since we need to provide signing information.
+            continue
+        }
```
As mentioned above, this can't be supported anyway.

akien-mga commented on platform/android/java/editor/build.gradle:
```diff
@@ -0,0 +1,53 @@
+    ndkVersion versions.ndkVersion
+
+    defaultConfig {
+        // The 'applicationId' suffix allows to install Godot 3.x(x3) and 4.x(x4) on the same device
+        applicationId "org.godotengine.editor.x3"
```
Maybe `v3` and `v4` would be clearer? Assuming `3x` and `4x` were not possible due to restrictions on identifiers starting with a number?

akien-mga commented on platform/android/java/editor/build.gradle:
```diff
@@ -0,0 +1,53 @@
+        // The 'applicationId' suffix allows to install Godot 3.x(x3) and 4.x(x4) on the same device
+        applicationId "org.godotengine.editor.x3"
+        minSdkVersion versions.minSdk
+        //noinspection ExpiredTargetSdkVersion - Restrict to version 29 until https://github.com/godotengine/godot/pull/51815 is submitted
+        targetSdkVersion 29 // versions.targetSdk
```
I guess that means this could be published on Google Play until this is solved? (Not that it's a problem right now, we don't have explicit plans to do so at this time.)

m4gr3d replied:
Yes, that's correct. The current plan is to stick to target sdk 29 until we're done with alpha and beta releases for the editor.

akien-mga commented on platform/android/java/editor/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,70 @@
+    xmlns:tools="http://schemas.android.com/tools"
+    package="org.godotengine.editor"
+    android:installLocation="auto"
+    android:versionCode="1"
+    android:versionName="1.0">
```
Should those reflect the actual version from `version.py`?

akien-mga commented on platform/android/java/editor/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,70 @@
+            android:resizeableActivity="false"
+            android:screenOrientation="landscape"
+            android:theme="@android:style/Theme.Black.NoTitleBar.Fullscreen">
+        </activity>
+
+
+    </application>
```
Style nitpick (keeping only one empty line instead of two):
```suggestion
        </activity>

    </application>
```

akien-mga commented on platform/android/java/lib/build.gradle:
```diff
@@ -49,24 +60,39 @@ android {
+        // Tools enabled jni library
+        toolsEnabledDebug.jniLibs.srcDirs = ['libs/tools/debug']
+        toolsEnabledRelease_debug.jniLibs.srcDirs = ['libs/tools/release_debug']
+        toolsEnabledRelease.jniLibs.srcDirs = ['libs/tools/release']
```
This one is not supported.

m4gr3d replied:
Removed.

akien-mga commented on platform/android/java/editor/src/release_debug/res/values/strings.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+	<string name="godot_editor_name_string">Godot Editor (release-debug)</string>
```
It's probably best not to introduce another naming discrepancy here:
```suggestion
	<string name="godot_editor_name_string">Godot Editor (release_debug)</string>
```

What's the difference between this and the one in `editor/src/main` though?

m4gr3d replied:
Since the `release` target is not valid for the editor build, this is the same as `editor/src/main` so I removed the `release_debug` subdirectory (including this file).

m4gr3d replied:
The resulting naming convention is `Godot Editor` for the `release_debug` builds and `Godot Editor (debug)` for the `debug` builds.

akien-mga commented on platform/android/java/lib/build.gradle:
```diff
@@ -101,14 +127,20 @@ android {
+            def taskName = getSconsTaskName(flavorName, buildType, selectedAbi)
+            tasks.create(name: taskName, type: Exec) {
+                executable sconsExecutableFile.absolutePath
+                args "tools=${toolsFlag}", "--directory=${pathToRootDir}", "platform=android", "target=${releaseTarget}", "android_arch=${selectedAbi}", "-j" + Runtime.runtime.availableProcessors()
```
Nitpick but I'd put `tools` after `platform`, that's how we usually document it (it does work in any order of course).

akien-mga commented on platform/android/java/nativeSrcsConfigs/CMakeLists.txt:
```diff
@@ -17,3 +17,5 @@ target_include_directories(${PROJECT_NAME}
         ${GODOT_ROOT_DIR}
         ${GODOT_ROOT_DIR}/modules/gdnative/include)
+
+add_definitions(-DUNIX_ENABLED -DTOOLS_ENABLED)
```
That looks weird, but given the comment at the top I guess this is just added to allow Android Studio to consider code in those `#ifdef`s as part of the build config?

m4gr3d replied:
Yep!

akien-mga commented on platform/android/os_android.cpp:
```diff
@@ -279,6 +292,15 @@ void OS_Android::set_keep_screen_on(bool p_enabled) {
 }
 
+void OS_Android::set_low_processor_usage_mode(bool p_enabled) {
+#ifdef TOOLS_ENABLED
+	// Disabled as it causes flickers. We also expect the devices running Godot in editor mode to be high end.
+	OS::set_low_processor_usage_mode(false);
```
This is something that should be looked into eventually. Many users want to use low processor usage mode on Android and are having issues with it - and I do think it could be useful for the editor too, even on high end. You don't want the battery to be drained just because you had the Godot editor open to read its built-in documentation.

Should likely also call `OS_Unix` implementation in case it gets overridden in the future.

akien-mga commented on platform/android/os_android.cpp:
```diff
@@ -313,8 +335,13 @@ bool OS_Android::main_loop_iterate() {
 void OS_Android::main_loop_end() {
-	if (main_loop)
+	if (main_loop) {
+		if (main_loop->is_class("SceneTree")) {
+			SceneTree *scene_tree = Object::cast_to<SceneTree>(main_loop);
+			scene_tree->quit();
+		}
```
Skip the unnecessary string check:
```suggestion
		SceneTree *scene_tree = Object::cast_to<SceneTree>(main_loop);
		if (scene_tree) {
			scene_tree->quit();
		}
```

akien-mga commented on platform/android/os_android.cpp:
```diff
@@ -520,6 +555,10 @@ void OS_Android::vibrate_handheld(int p_duration_ms) {
 }
 
+String OS_Android::get_config_path() const {
+	return get_user_data_dir() + "/config";
```
```suggestion
	return get_user_data_dir().plus_file("config");
```

akien-mga commented on platform/android/os_android.cpp:
```diff
@@ -448,6 +479,10 @@ String OS_Android::get_data_path() const {
 }
 
+String OS_Android::get_executable_path() const {
+	return OS::get_executable_path();
```
Is it intentional to bypass `OS_Unix::get_executable_path()` by going straight to `OS`?
If so, this warrants a commit to clarify (otherwise it looks like either a redundant override or a mistake).

m4gr3d replied:
Yes, this is done so we can detect when we're running on Android, in which case `apk` is returned as the executable path, and act accordingly.
The unix logic returns the unix process path whose access can be limited depending on which version of Android the device is running.

akien-mga commented on platform/android/os_android.cpp:
```diff
@@ -393,7 +420,11 @@ Error OS_Android::shell_open(String p_uri) {
 
 String OS_Android::get_resource_dir() const {
+#ifdef TOOLS_ENABLED
+	return OS::get_resource_dir();
```
It's not defined currently but it might be safer to use `OS_Unix::get_resource_dir()`?

Commit:
Cleanup the gradle build configuration

These changes resolve the inconsistency between gradle and scons targets by configuring the gradle buildtypes to match the scons targets.

thebestnom reviewed (commented):
I think that only squashing the commits, the changes to my old code looks ok, Ill test it again on my phone and approve

szymonm-google reviewed (commented):
@m4gr3d, I went over the changes and didn't find any issues, but since I'm not that familiar with the code base, I don't feel empowered to lgtm it.

m4gr3d commented:
> @m4gr3d, I went over the changes and didn't find any issues, but since I'm not that familiar with the code base, I don't feel empowered to lgtm it.

@szymonm-google Thanks for taking a look! Hopefully we can quickly get you up to speed with the code base :)

akien-mga commented on platform/android/java/editor/src/debug/res/values/strings.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+	<string name="godot_editor_name_string">Godot Editor (debug)</string>
```
This is used by the `target=release_debug` build, which is the "production" build target for the editor. So it's a bit weird to have the installed application be called `Godot Editor (debug)`, it should be just `Godot Editor` and `(debug)` should be added only for `target=debug` builds.

akien-mga reviewed (approved):
Tested on a Xiaomi Pocophone F1, it seems to work great! No issue building the APK from Linux either with the provided instructions.
I could install the 2D platformer from the Asset Library and run it, with touchscreen controls already supported (that's luck, it's one of the few demos that supports it :P).

Seems good to merge, and we can iterate further in new issues/PRs.

Some ideas/remarks for future work:
- It would be good if the Back button when running a project from the editor could close the project and go back to the editor (currently it closes everything). It's already handled in the editor itself to prevent it from closing the editor, but not at runtime.
- We often have pre-selected filter text fields in the UI, and on Android (at least without a bluetooth keyboard) that shows the virtual keyboard. It confirms that the virtual keyboard works well and is properly called when in a LineEdit, but here we might want to figure out a way to prevent it from showing automatically on such pre-selected fields, and only show it when the user actually clicks the field.

akien-mga commented:
Thanks @thebestnom for an amazing port and @m4gr3d for the second pass!

LightningAA commented:
Maybe after a few more of the issues are worked out, but what would it take to get this on the Google Play store (and possibly other app stores)?

thebestnom commented:
And thanks again for @fahrstuhl for first trying to do it 😄

akien-mga commented:
> Maybe after a few more of the issues are worked out, but what would it take to get this on the Google Play store (and possibly other app stores)?

That's planned further down the line :)

dugramen commented:
This is amazing! I have suggestion for touch use though. It's currently impossible to right click anything, I think holding down and releasing something should be considered a right click, as that's how it's usually done on mobile.

Also dragging should only occur if the Control was held for a short period before moving. Otherwise it's difficult to scroll. 

Basically: 
Hold down and release -> Right Click
Hold down and move -> Drag

dugramen commented:
Another thing I noticed, which may not be related to this but to text fields on mobile in general. Usually if you tap the text field, a handle appears, and if you click it you get a menu for copy, paste etc. That's not here for now. Maybe there's a way to let Android itself handle the text field?

Calinou commented:
> Maybe there's a way to let Android itself handle the text field?

This already happens with an invisible text field control, but we can't replace all LineEdits/TextEdits with native Android controls as they don't support all the features that LineEdit and TextEdit provide.

dugramen commented:
> > Maybe there's a way to let Android itself handle the text field?
> 
> This already happens with an invisible text field control, but we can't replace all LineEdits/TextEdits with native Android controls as they don't support all the features that LineEdit and TextEdit provide.

I see. How about just for the script editor then, as I feel copy and paste is pretty important there (As well as precise cursor movement when dragging the space bar)

Calinou commented:
> How about just for the script editor then, as I feel copy and paste is pretty important there (As well as precise cursor movement when dragging the space bar)

The script editor is literally the worst candidate for using a native Android control :slightly_smiling_face: 

The only way to implement this is to implement long-tapping selection in LineEdit and TextEdit. This is a lot of work to do correctly, and I'm not sure if it can be done at all.

akien-mga commented:
We'll be gathering feedback on the Android editor in this proposal: https://github.com/godotengine/godot-proposals/issues/3931

Feel free to add your suggestions there. If you find actual bugs, you can of course open bug reports like usual.

dugramen commented:
> The script editor is literally the worst candidate for using a native Android control, since it's so advanced 🙂
> 
> The only way to implement this is to implement long-tapping selection in LineEdit and TextEdit. This is a lot of work to do correctly, and I'm not sure if it can be done at all. The same applies to dragging the spacebar, but I don't know how virtual keyboards actually translate this to the application. Do they send keyboard "left arrow" and "right arrow" events, or do something special? If this is something special, can C++ applications intercept it somehow?

Oh I assumed Android controls could be added on top of everything else, but I guess not 😅. I remembered seeing code editors in Android before, but after looking at how they do it, it seems like a lot of these things they had to implement themselves. But they WERE able to get rid of the suggestion bar at least

https://imgur.com/a/AdYDPVG
Top bar: Select all, cut, copy, paste
Bottom bar: Tab, undo, redo, search, left_key, right_key, save

alex9310 commented:
I just didn’t understand where can I get android sdk to build apk and aab?

akien-mga commented:
> I just didn’t understand where can I get android sdk to build apk and aab?

See https://docs.godotengine.org/en/stable/development/compiling/compiling_for_android.html for documentation.

Starkium commented:
this runs on an oculus quest for those who were wondering :)

I'd like to get a way to launch into vr mode from it or even start the editor in a vr mode. This is going to be crazy for vr development.

alex9310 commented:
> > I just didn’t understand where can I get android sdk to build apk and aab?
> 
> See https://docs.godotengine.org/en/stable/development/compiling/compiling_for_android.html for documentation.

as I understand it, I collect it on the pc as usual, did I understand correctly?

2439905184 commented:
> `3.x` version of the Godot Editor Android port based on the work done by @thebestnom in #36776!
> 
> This PR cleans a lot of the logic in #36776 and adapt the code for the `3.x` branch. Given the stability of the `3.x` branch, this should help accelerate testing and feedback on this implementation.
> 
> There are still a lot of tweaks, UI cleanup and touch-ups needed to be done before the Editor Android port can be considered suitable for public release and usage. Nonetheless, the current logic is stable enough and partitioned away that we can start reviewing and landing this PR and then iterate on top of it.
> 
> Addresses [godotengine/godot-proposals#3931](https://github.com/godotengine/godot-proposals/issues/3931)
> 
>  godot_android_editor_on_device_run_through.mp4 
> ### Apk Download (only meant for testing and feedback!)
> [Latest 3.x Godot Editor Android build (all architectures)](https://drive.google.com/drive/folders/1Tzhu3yZFnOKwmLUCRBdbMMMHn79e_MmL?usp=sharing)
> 
> ### Generate apk from sources
> * Using the regular Godot build process, generate Godot tools android shared libraries for each platform you want the apk to support:
>   
>   * e.g: `scons tools=true platform=android target=<build_target> android_arch=<android_arch>`
>   * `release_debug` is recommended for the `target` option
>   * `android_arch` should be one of `arm64v8`, `armv7`, `x86` or `x86_64`
> * Navigate to the `platform/android/java` directory
> * Run `./gradlew generateGodotEditor`
>   
>   * The command will generate one apk for each available target
>   * The generated apk(s) can be found in the `bin` directory.
mouse right click doesn't work.
Device: Android8.0 huawei with bluetooth mouse and keyboard

thebestnom commented:
> mouse right click doesn't work.\nDevice

In most stock rom Android recognize right click as back button in a non configureable way... Only fix is to replace rom 😅 (also middle click as home which is so frustrating)

2439905184 commented:
> 

maybe just add button "del node" and then just click it to remove nodes in editor.
Because right click doesn't work .

badduck32 commented:
Editing scripts is very hard, I propose that there should be a way to edit scripts in portrait mode. I used an app to force portrait mode on the Godot editor, and it looks very good, apart from the left panel which I don't know how to turn off.
If this were full-screen I feel like it would be very usable. Maybe there should be a setting to make the entire editor portrait mode?

![Screenshot_20220430-164842](https://user-images.githubusercontent.com/44337504/166110471-bfd24a27-74f0-47e2-812d-25cb9c46032e.jpg)

The app looks and feels great in general, it hasn't crashed for me and it runs my project well, I'm excited for the future!!!

Starkium commented:
I'd love virtual windows for editing in vr

On Sat, Apr 30, 2022, 10:54 AM Badduck ***@***.***> wrote:

> Editing scripts is very hard, I propose that there should be a way to edit
> scripts in portrait mode. I used an app to force portrait mode on the Godot
> editor, and it looks very good, apart from the left panel which I don't
> know how to turn off.
> If this were full-screen I feel like it would be very usable. Maybe there
> should be a setting to make the entire editor portrait mode?
>
> [image: Screenshot_20220430-164842]
> <https://user-images.githubusercontent.com/44337504/166110471-bfd24a27-74f0-47e2-812d-25cb9c46032e.jpg>
>
> The app looks and feels great in general, it hasn't crashed for me and it
> runs my project well, I'm excited for the future!!!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/57747#issuecomment-1114001545>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAVTN7CVGW6BLVXLRVWU5DVHVCSNANCNFSM5NXCLB2Q>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Calinou commented:
> I used an app to force portrait mode on the Godot editor, and it looks very good, apart from the left panel which I don't know how to turn off.

You can enable distraction-free mode by clicking the "4 outward arrows" icon in the top-right corner of the 2D/3D/script tab:

![image](https://user-images.githubusercontent.com/180032/166111600-4c6021a6-577d-48ec-83f6-c7123b053d6f.png)

If you have a hardware keyboard, you can also press <kbd>Ctrl + Shift + F11</kbd> to toggle it.

badduck32 commented:
> Hiding the script list is done by clicking the leftwards arrow at the bottom of the script editor (it's obstructed by the virtual keyboard on your screenshot):

That works like a charm! unfortunately you can't scroll using only touch input as of yet, and the scroll bar is off-screen when I force portrait mode, so there will have to be some slight changes but it's looking very promising!

Pedrokkkkkk commented:
Hello, i made a plugin that helps using it on android, i think its not relevant to show here, btw the plugin handles touch inputs and converts into mouse inputs, a keyboard inside it is also included, link to the project at github there:

[Godot Touch Acessibility Plugin](https://github.com/Pedrokkkkkk/Godot-Touch-Acessibility/releases/tag/v1.0.0)

Hanprogramer commented:
![Screenshot_2022-05-16-00-00-13-788_org godotengine editor v3](https://user-images.githubusercontent.com/30620423/168484837-c14d2153-c53c-485b-bc7d-211bfc500ef7.jpg)
I cant access some parts on the bottom of the screen probably because my phone has a camera notch on the screen (Redmi note 10 pro)

Calinou commented:
> I cant access some parts on the bottom of the screen probably because my phone has a camera notch on the screen (Redmi note 10 pro)

This is not related to your device having notch. It's just that there isn't enough vertical space to fit the Animation bottom panel (which has a minimum height).

Jake012345 commented:
> Hello, i made a plugin that helps using it on android, i think its not relevant to show here, btw the plugin handles touch inputs and converts into mouse inputs, a keyboard inside it is also included, link to the project at github there:
> 
> [Godot Touch Acessibility Plugin](https://github.com/Pedrokkkkkk/Godot-Touch-Acessibility/releases/tag/v1.0.0) 

Is nice but how do i use it? 😬

Jake012345 commented:
> > Hello, i made a plugin that helps using it on android, i think its not relevant to show here, btw the plugin handles touch inputs and converts into mouse inputs, a keyboard inside it is also included, link to the project at github there:
> > 
> > [Godot Touch Acessibility Plugin](https://github.com/Pedrokkkkkk/Godot-Touch-Acessibility/releases/tag/v1.0.0) 
> 
> Is nice but how do i use it? 😬

I see. Thats pretty helpful. Thanks for your work 😊
